File: android/src/main/java/com/cipace/capacitor/videorecorder/VideoRecorderPlugin.java
package com.cipace.capacitor.videorecorder;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
    name = "CipaceVideoRecorder",
    permissions = {
        @Permission(
            strings = { Manifest.permission.CAMERA },
            alias = VideoRecorderPlugin.CAMERA
        ),
        @Permission(
            strings = { Manifest.permission.RECORD_AUDIO },
            alias = VideoRecorderPlugin.MICROPHONE
        ),
        @Permission(
            strings = { Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE },
            alias = VideoRecorderPlugin.STORAGE
        ),
        @Permission(
            strings = { "android.permission.READ_MEDIA_VIDEO" },
            alias = VideoRecorderPlugin.MEDIA_VIDEO
        )
    }
)
public class VideoRecorderPlugin extends Plugin {
    
    static final String CAMERA = "camera";
    static final String MICROPHONE = "microphone";
    static final String STORAGE = "storage";
    static final String MEDIA_VIDEO = "mediaVideo";

    private VideoRecorder videoRecorder;
    @Override
    public void load() {
        videoRecorder = new VideoRecorder(getContext());
    }
    
    // MARK: - Media Capture Compatible Methods
    
    @PluginMethod
    public void captureVideo(PluginCall call) {
        // 简化：像 cordova-plugin-media-capture 一样，直接启动录制
        // 让 VideoRecordingActivity 自己处理权限请求
        VideoRecordingOptions options = createOptionsFromCall(call);
        Intent intent = new Intent(getContext(), VideoRecordingActivity.class);
        intent.putExtra(VideoRecordingActivity.EXTRA_OPTIONS, options);
        intent.putExtra(VideoRecordingActivity.EXTRA_IS_CAPTURE_MODE, true);
        startActivityForResult(call, intent, "handleCaptureVideoResult");
    }

    @PermissionCallback
    private void captureVideoWithPermissions(PluginCall call) {
        if (!hasFullPermissions()) {
            call.reject("PERMISSION_DENIED", "Camera, microphone and storage permissions are required for video recording");
            return;
        }

        VideoRecordingOptions options = createOptionsFromCall(call);

        Intent intent = new Intent(getContext(), VideoRecordingActivity.class);
        intent.putExtra(VideoRecordingActivity.EXTRA_OPTIONS, options);
        intent.putExtra(VideoRecordingActivity.EXTRA_IS_CAPTURE_MODE, true);

        startActivityForResult(call, intent, "handleCaptureVideoResult");
    }

    @ActivityCallback
    private void handleCaptureVideoResult(PluginCall call, androidx.activity.result.ActivityResult result) {
        if (result.getResultCode() != Activity.RESULT_OK || result.getData() == null) {
            call.reject("USER_CANCELLED", "Recording was cancelled");
            return;
        }

        Intent data = result.getData();

        VideoRecorderError error = (VideoRecorderError) data.getSerializableExtra(VideoRecordingActivity.EXTRA_ERROR);
        if (error != null) {
            call.reject(error.code, error.message, error.details);
            return;
        }

        VideoRecorder.StopRecordingResult recordingResult = (VideoRecorder.StopRecordingResult) data.getSerializableExtra(VideoRecordingActivity.EXTRA_RESULT);
        if (recordingResult != null) {
            JSObject mediaFile = new JSObject();
            mediaFile.put("name", new File(recordingResult.videoPath).getName());
            mediaFile.put("fullPath", recordingResult.videoPath);
            mediaFile.put("type", recordingResult.mimeType);
            mediaFile.put("lastModifiedDate", recordingResult.endTime);
            mediaFile.put("size", recordingResult.fileSize);

            JSArray files = new JSArray();
            files.put(mediaFile);

            JSObject ret = new JSObject();
            ret.put("files", files);
            call.resolve(ret);
        } else {
            call.reject("UNKNOWN_ERROR", "Unknown error occurred during recording");
        }
    }
    
    @PluginMethod
    public void captureAudio(PluginCall call) {
        if (getPermissionState(Manifest.permission.RECORD_AUDIO) != PermissionState.GRANTED) {
            requestPermissionForAlias(MICROPHONE, call, "captureAudioWithPermissions");
            return;
        }

        captureAudioWithPermissions(call);
    }
    
    @PermissionCallback
    private void captureAudioWithPermissions(PluginCall call) {
        if (getPermissionState(Manifest.permission.RECORD_AUDIO) != PermissionState.GRANTED) {
            call.reject("PERMISSION_DENIED", "Microphone permission denied");
            return;
        }
        
        VideoRecordingOptions options = createOptionsFromCall(call);
        
        videoRecorder.captureAudio(options, new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                if (result instanceof VideoRecorder.CaptureAudioResult) {
                    VideoRecorder.CaptureAudioResult captureResult = (VideoRecorder.CaptureAudioResult) result;
                    JSObject ret = new JSObject();
                    ret.put("files", captureResult.files);
                    call.resolve(ret);
                }
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }
    
    @PluginMethod
    public void getSupportedVideoModes(PluginCall call) {
        JSObject[] supportedModes = {
            createModeObject("video/mp4", 640, 480),
            createModeObject("video/mp4", 1280, 720),
            createModeObject("video/mp4", 1920, 1080),
            createModeObject("video/mp4", 3840, 2160)
        };
        call.resolve(new JSObject().put("modes", supportedModes));
    }
    
    @PluginMethod
    public void getSupportedAudioModes(PluginCall call) {
        JSObject[] supportedModes = {
            createModeObject("audio/mp4", 0, 0),
            createModeObject("audio/wav", 0, 0),
            createModeObject("audio/aac", 0, 0)
        };
        call.resolve(new JSObject().put("modes", supportedModes));
    }
    
    // MARK: - Advanced Recording Methods
    
    @PluginMethod
    public void openRecordingInterface(PluginCall call) {
        if (getPermissionState(Manifest.permission.CAMERA) != PermissionState.GRANTED) {
            requestPermissionForAlias(CAMERA, call, "openRecordingInterfaceWithPermissions");
            return;
        }

        openRecordingInterfaceWithPermissions(call);
    }
    
    @PermissionCallback
    private void openRecordingInterfaceWithPermissions(PluginCall call) {
        if (getPermissionState(Manifest.permission.CAMERA) != PermissionState.GRANTED) {
            call.reject("PERMISSION_DENIED", "Camera permission denied");
            return;
        }

        VideoRecordingOptions options = createOptionsFromCall(call);

        Intent intent = new Intent(getContext(), VideoRecordingActivity.class);
        intent.putExtra(VideoRecordingActivity.EXTRA_OPTIONS, options);

        startActivityForResult(call, intent, "handleRecordingResult");
    }

    @ActivityCallback
    private void handleRecordingResult(PluginCall call, Intent data) {
        if (data == null) {
            call.reject("USER_CANCELLED", "Recording was cancelled");
            return;
        }

        VideoRecorderError error = (VideoRecorderError) data.getSerializableExtra(VideoRecordingActivity.EXTRA_ERROR);
        if (error != null) {
            call.reject(error.code, error.message, error.details);
            return;
        }

        VideoRecorder.StopRecordingResult result = (VideoRecorder.StopRecordingResult) data.getSerializableExtra(VideoRecordingActivity.EXTRA_RESULT);
        if (result != null) {
            JSObject ret = new JSObject();
            ret.put("recordingId", result.recordingId);
            ret.put("videoPath", result.videoPath);
            ret.put("fileSize", result.fileSize);
            ret.put("duration", result.duration);
            ret.put("width", result.width);
            ret.put("height", result.height);
            ret.put("startTime", result.startTime);
            ret.put("endTime", result.endTime);
            ret.put("thumbnailPath", result.thumbnailPath);
            ret.put("mimeType", result.mimeType);

            call.resolve(ret);
        } else {
            call.reject("UNKNOWN_ERROR", "Unknown error occurred during recording");
        }
    }

    
    @PluginMethod
    public void stopRecording(PluginCall call) {
        videoRecorder.stopRecording(new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                if (result instanceof VideoRecorder.StopRecordingResult) {
                    VideoRecorder.StopRecordingResult stopResult = (VideoRecorder.StopRecordingResult) result;
                    JSObject ret = new JSObject();
                    ret.put("recordingId", stopResult.recordingId);
                    ret.put("videoPath", stopResult.videoPath);
                    ret.put("fileSize", stopResult.fileSize);
                    ret.put("duration", stopResult.duration);
                    ret.put("width", stopResult.width);
                    ret.put("height", stopResult.height);
                    ret.put("startTime", stopResult.startTime);
                    ret.put("endTime", stopResult.endTime);
                    ret.put("thumbnailPath", stopResult.thumbnailPath);
                    ret.put("mimeType", stopResult.mimeType);
                    call.resolve(ret);
                }
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }
    
    @PluginMethod
    public void pauseRecording(PluginCall call) {
        videoRecorder.pauseRecording(new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                call.resolve();
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }
    
    @PluginMethod
    public void resumeRecording(PluginCall call) {
        videoRecorder.resumeRecording(new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                call.resolve();
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }
    
    @PluginMethod
    public void checkPermissions(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("camera", getPermissionState(Manifest.permission.CAMERA));
        ret.put("microphone", getPermissionState(Manifest.permission.RECORD_AUDIO));
        ret.put("storage", getStoragePermissionState());
        call.resolve(ret);
    }
    
    @PluginMethod
    public void requestPermissions(PluginCall call) {
        // 简化：直接返回当前权限状态
        checkPermissions(call);
    }
    
    @PluginMethod
    public void getRecordingStatus(PluginCall call) {
        VideoRecorder.RecordingStatus status = videoRecorder.getRecordingStatus();
        JSObject ret = new JSObject();
        ret.put("isRecording", status.isRecording);
        ret.put("isPaused", status.isPaused);
        ret.put("currentDuration", status.currentDuration);
        ret.put("recordingId", status.recordingId);
        call.resolve(ret);
    }
    
    @PluginMethod
    public void deleteRecording(PluginCall call) {
        String videoPath = call.getString("videoPath");
        boolean deleteThumbnail = call.getBoolean("deleteThumbnail", true);
        
        if (videoPath == null) {
            call.reject("INVALID_OPTIONS", "videoPath is required");
            return;
        }
        
        VideoRecorder.deleteRecording(videoPath, deleteThumbnail, new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                call.resolve();
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }
    
    @PluginMethod
    public void generateThumbnail(PluginCall call) {
        String videoPath = call.getString("videoPath");
        double timeAt = call.getDouble("timeAt", 1.0); // 默认在第1秒生成缩略图
        double quality = call.getDouble("quality", 0.8); // 默认压缩质量0.8
        
        if (videoPath == null) {
            call.reject("INVALID_OPTIONS", "videoPath is required");
            return;
        }
        
        // 处理路径兼容性：支持 file:// 开头的路径
        String actualVideoPath;
        if (videoPath.startsWith("file://")) {
            actualVideoPath = videoPath.substring(7); // 移除 "file://" 前缀
        } else {
            actualVideoPath = videoPath;
        }
        
        // 验证视频文件是否存在
        java.io.File videoFile = new java.io.File(actualVideoPath);
        if (!videoFile.exists()) {
            call.reject("FILE_NOT_FOUND", "Video file not found at path: " + actualVideoPath);
            return;
        }
        
        VideoRecorder.generateThumbnail(actualVideoPath, timeAt, quality, new VideoRecorder.RecordingCallback() {
            @Override
            public void onSuccess(Object result) {
                if (result instanceof VideoRecorder.ThumbnailResult) {
                    VideoRecorder.ThumbnailResult thumbnailResult = (VideoRecorder.ThumbnailResult) result;
                    JSObject ret = new JSObject();
                    ret.put("thumbnailPath", thumbnailResult.thumbnailPath);
                    ret.put("videoPath", actualVideoPath);
                    ret.put("timeAt", timeAt);
                    ret.put("quality", quality);
                    call.resolve(ret);
                }
            }
            
            @Override
            public void onError(VideoRecorderError error) {
                call.reject(error.code, error.message, error.details);
            }
        });
    }

    // MARK: - Helper Methods

    private VideoRecordingOptions createOptionsFromCall(PluginCall call) {
        VideoRecordingOptions options = new VideoRecordingOptions();
        options.quality = call.getString("quality", "high");
        options.maxDuration = call.getDouble("maxDuration", 300.0);
        options.fileNamePrefix = call.getString("fileNamePrefix", "video_recording");
        options.saveToGallery = call.getBoolean("saveToGallery", false);
        options.camera = call.getString("camera", "back");
        options.orientation = call.getString("orientation", "portrait");
        options.enableAudio = call.getBoolean("enableAudio", true);

        Double duration = call.getDouble("duration");
        if (duration != null) {
            options.maxDuration = duration;
        }

        Integer qualityNumber = call.getInt("quality");
        if (qualityNumber != null) {
            options.quality = mapQualityFromNumber(qualityNumber);
        }

        return options;
    }

    private String mapQualityFromNumber(int quality) {
        if (quality <= 25) return "low";
        if (quality <= 50) return "medium";
        if (quality <= 75) return "high";
        return "highest";
    }

    private JSObject createModeObject(String type, int width, int height) {
        JSObject mode = new JSObject();
        mode.put("type", type);
        mode.put("width", width);
        mode.put("height", height);
        return mode;
    }

    @Override
    public boolean hasRequiredPermissions() {
        return hasVideoPermissions();
    }

    private boolean hasVideoPermissions() {
        return getPermissionState(Manifest.permission.CAMERA) == PermissionState.GRANTED &&
               hasStoragePermission();
    }

    private boolean hasAudioPermissions() {
        return getPermissionState(Manifest.permission.RECORD_AUDIO) == PermissionState.GRANTED &&
               hasStoragePermission();
    }

    private boolean hasFullPermissions() {
        return getPermissionState(Manifest.permission.CAMERA) == PermissionState.GRANTED &&
               getPermissionState(Manifest.permission.RECORD_AUDIO) == PermissionState.GRANTED &&
               hasStoragePermission();
    }





    private boolean hasStoragePermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Android 13+ 使用新的媒体权限
            PermissionState videoPermission = getPermissionState("android.permission.READ_MEDIA_VIDEO");
            android.util.Log.d("VideoRecorder", "Android 13+ 视频权限状态: " + videoPermission);
            return videoPermission == PermissionState.GRANTED;
        } else {
            // Android 12 及以下使用传统存储权限
            PermissionState writePermission = getPermissionState(Manifest.permission.WRITE_EXTERNAL_STORAGE);
            PermissionState readPermission = getPermissionState(Manifest.permission.READ_EXTERNAL_STORAGE);
            android.util.Log.d("VideoRecorder", "传统存储权限 - Write: " + writePermission + ", Read: " + readPermission);
            return writePermission == PermissionState.GRANTED && readPermission == PermissionState.GRANTED;
        }
    }

    private String getStoragePermissionState() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return getPermissionState("android.permission.READ_MEDIA_VIDEO").toString().toLowerCase();
        } else {
            return getPermissionState(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PermissionState.GRANTED &&
                   getPermissionState(Manifest.permission.READ_EXTERNAL_STORAGE) == PermissionState.GRANTED
                   ? "granted" : "denied";
        }
    }


}

File: ios/Sources/VideoRecorderPlugin/VideoRecorderPlugin.swift
import Foundation
import Capacitor
import AVFoundation
import Photos
import UIKit

@objc(VideoRecorderPlugin)
public class VideoRecorderPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VideoRecorderPlugin"
    public let jsName = "CipaceVideoRecorder"
    public let pluginMethods: [CAPPluginMethod] = [
        // Media Capture 兼容方法
        CAPPluginMethod(name: "captureVideo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "captureAudio", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSupportedVideoModes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSupportedAudioModes", returnType: CAPPluginReturnPromise),
        
        // 高级录制方法
        CAPPluginMethod(name: "openRecordingInterface", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRecordingStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteRecording", returnType: CAPPluginReturnPromise),
        
        // 缩略图生成方法
        CAPPluginMethod(name: "generateThumbnail", returnType: CAPPluginReturnPromise)
    ]
    
    private var videoRecorder: VideoRecorder?
    private var currentCall: CAPPluginCall?
    private var recordingInterfaceCall: CAPPluginCall?
    
    // MARK: - Media Capture 兼容方法
    
    @objc func captureVideo(_ call: CAPPluginCall) {
        let duration = call.getDouble("duration") ?? 300.0
        let limit = call.getInt("limit") ?? 1
        let quality = call.getInt("quality") ?? 50

        // 转换为内部选项格式
        let options = VideoRecordingOptions(
            quality: mapQualityFromNumber(quality),
            maxDuration: duration,
            fileNamePrefix: "captured_video",
            saveToGallery: false,
            camera: .back,
            orientation: .portrait,
            enableAudio: true
        )

        videoRecorder = VideoRecorder()

        // 存储call以便在录制完成时使用
        self.currentCall = call

        videoRecorder?.openRecordingInterface(options: options) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let recordingResult):
                    // 录制完成，直接返回结果
                    print("Recording completed with ID: \(recordingResult.recordingId)")

                    // 检查是否有来自captureVideo的待处理调用
                    if let captureCall = self?.currentCall {
                        // 这是captureVideo调用的结果，返回MediaCapture格式
                        let mediaFile = self?.createMediaFileDict(from: recordingResult)
                        captureCall.resolve(["files": [mediaFile].compactMap { $0 }])
                        self?.currentCall = nil
                    } else {
                        // 这是直接的openRecordingInterface调用，返回完整的录制结果
                        let response = self?.createStopRecordingResponse(from: recordingResult)
                        call.resolve(response ?? JSObject())
                    }
                    self?.videoRecorder = nil
                case .failure(let error):
                    // 检查是否有待处理的调用
                    if let captureCall = self?.currentCall {
                        captureCall.reject(error.code, error.message, nil, error.details)
                        self?.currentCall = nil
                    } else {
                        call.reject(error.code, error.message, nil, error.details)
                    }
                }
            }
        }
    }
    
    @objc func captureAudio(_ call: CAPPluginCall) {
        let duration = call.getDouble("duration") ?? 300.0
        let limit = call.getInt("limit") ?? 1
        
        // 使用系统的音频录制功能
        let audioRecorder = AudioRecorder()
        
        audioRecorder.startRecording(duration: duration) { result in
            DispatchQueue.main.async {
                switch result {
                case .success(let audioFiles):
                    let files = audioFiles.map { self.createMediaFileDict(from: $0) }
                    call.resolve(["files": files])
                case .failure(let error):
                    call.reject("RECORDING_FAILED", error.localizedDescription)
                }
            }
        }
    }
    
    @objc func getSupportedVideoModes(_ call: CAPPluginCall) {
        let supportedModes = [
            ["type": "video/mp4", "width": 640, "height": 480],
            ["type": "video/mp4", "width": 1280, "height": 720],
            ["type": "video/mp4", "width": 1920, "height": 1080],
            ["type": "video/mp4", "width": 3840, "height": 2160]
        ]
        call.resolve(["supportedModes": supportedModes])
    }
    
    @objc func getSupportedAudioModes(_ call: CAPPluginCall) {
        let supportedModes = [
            ["type": "audio/mp4", "width": 0, "height": 0],
            ["type": "audio/wav", "width": 0, "height": 0],
            ["type": "audio/aac", "width": 0, "height": 0]
        ]
        call.resolve(["supportedModes": supportedModes])
    }
    
    // MARK: - 高级录制方法
    
    @objc func openRecordingInterface(_ call: CAPPluginCall) {
        let quality = call.getString("quality") ?? "high"
        let maxDuration = call.getDouble("maxDuration") ?? 300.0
        let fileNamePrefix = call.getString("fileNamePrefix") ?? "video_recording"
        let saveToGallery = call.getBool("saveToGallery") ?? false
        let camera = call.getString("camera") ?? "back"
        let orientation = call.getString("orientation") ?? "portrait"
        let enableAudio = call.getBool("enableAudio") ?? true
        
        let options = VideoRecordingOptions(
            quality: VideoQuality(rawValue: quality) ?? .high,
            maxDuration: maxDuration,
            fileNamePrefix: fileNamePrefix,
            saveToGallery: saveToGallery,
            camera: CameraPosition(rawValue: camera) ?? .back,
            orientation: VideoOrientation(rawValue: orientation) ?? .portrait,
            enableAudio: enableAudio
        )
        
        videoRecorder = VideoRecorder()
        
        // 打开录制界面，等待录制完成后返回结果
        videoRecorder?.openRecordingInterface(options: options) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let recordingResult):
                    // 录制完成，返回完整结果
                    let response = [
                        "recordingId": recordingResult.recordingId,
                        "videoPath": recordingResult.videoPath,
                        "fileSize": recordingResult.fileSize,
                        "duration": recordingResult.duration,
                        "width": recordingResult.width,
                        "height": recordingResult.height,
                        "startTime": recordingResult.startTime,
                        "endTime": recordingResult.endTime,
                        "thumbnailPath": recordingResult.thumbnailPath ?? "",
                        "mimeType": recordingResult.mimeType
                    ] as [String : Any]

                    call.resolve(response)
                    self?.videoRecorder = nil
                case .failure(let error):
                    // 录制失败或被取消
                    call.reject(error.code, error.message, nil, error.details)
                    self?.videoRecorder = nil
                }
            }
        }
    }
    
    @objc func stopRecording(_ call: CAPPluginCall) {
        guard let recorder = videoRecorder else {
            call.reject("NOT_RECORDING", "No active recording to stop")
            return
        }

        recorder.stopRecording { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let recordingResult):
                    // 检查是否有来自captureVideo的待处理调用
                    if let captureCall = self?.currentCall {
                        // 这是captureVideo调用的结果，返回MediaCapture格式
                        let mediaFile = self?.createMediaFileDict(from: recordingResult)
                        captureCall.resolve(["files": [mediaFile].compactMap { $0 }])
                        self?.currentCall = nil
                    } else {
                        // 这是直接的stopRecording调用
                        let response = self?.createStopRecordingResponse(from: recordingResult)
                        call.resolve(response ?? JSObject())
                    }
                    self?.videoRecorder = nil
                case .failure(let error):
                    // 检查是否有待处理的调用
                    if let captureCall = self?.currentCall {
                        captureCall.reject(error.code, error.message, nil, error.details)
                        self?.currentCall = nil
                    } else {
                        call.reject(error.code, error.message, nil, error.details)
                    }
                }
            }
        }
    }
    
    @objc func pauseRecording(_ call: CAPPluginCall) {
        guard let recorder = videoRecorder else {
            call.reject("NOT_RECORDING", "No active recording to pause")
            return
        }
        
        recorder.pauseRecording { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject(error.code, error.message, nil, error.details)
                }
            }
        }
    }
    
    @objc func resumeRecording(_ call: CAPPluginCall) {
        guard let recorder = videoRecorder else {
            call.reject("NOT_RECORDING", "No paused recording to resume")
            return
        }
        
        recorder.resumeRecording { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject(error.code, error.message, nil, error.details)
                }
            }
        }
    }
    
    @objc public override func checkPermissions(_ call: CAPPluginCall) {
        let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
        let microphoneStatus = AVCaptureDevice.authorizationStatus(for: .audio)
        let photosStatus = PHPhotoLibrary.authorizationStatus()
        
        call.resolve([
            "camera": permissionStateString(from: cameraStatus),
            "microphone": permissionStateString(from: microphoneStatus),
            "storage": photosPermissionStateString(from: photosStatus)
        ])
    }
    
    @objc public override func requestPermissions(_ call: CAPPluginCall) {
        let group = DispatchGroup()
        var cameraGranted = false
        var microphoneGranted = false
        var photosGranted = false
        
        // 请求摄像头权限
        group.enter()
        AVCaptureDevice.requestAccess(for: .video) { granted in
            cameraGranted = granted
            group.leave()
        }
        
        // 请求麦克风权限
        group.enter()
        AVCaptureDevice.requestAccess(for: .audio) { granted in
            microphoneGranted = granted
            group.leave()
        }
        
        // 请求相册权限
        group.enter()
        PHPhotoLibrary.requestAuthorization { status in
            photosGranted = status == .authorized
            group.leave()
        }
        
        group.notify(queue: .main) {
            let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
            let microphoneStatus = AVCaptureDevice.authorizationStatus(for: .audio)
            let photosStatus = PHPhotoLibrary.authorizationStatus()
            
            call.resolve([
                "camera": self.permissionStateString(from: cameraStatus),
                "microphone": self.permissionStateString(from: microphoneStatus),
                "storage": self.photosPermissionStateString(from: photosStatus)
            ])
        }
    }
    
    @objc func getRecordingStatus(_ call: CAPPluginCall) {
        guard let recorder = videoRecorder else {
            call.resolve([
                "isRecording": false,
                "isPaused": false,
                "currentDuration": 0,
                "recordingId": ""
            ])
            return
        }
        
        let status = recorder.getRecordingStatus()
        call.resolve([
            "isRecording": status.isRecording,
            "isPaused": status.isPaused,
            "currentDuration": status.currentDuration,
            "recordingId": status.recordingId ?? ""
        ])
    }
    
    @objc func deleteRecording(_ call: CAPPluginCall) {
        guard let videoPath = call.getString("videoPath") else {
            call.reject("INVALID_OPTIONS", "videoPath is required")
            return
        }

        let deleteThumbnail = call.getBool("deleteThumbnail") ?? true

        VideoRecorder.deleteRecording(videoPath: videoPath, deleteThumbnail: deleteThumbnail) { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject(error.code, error.message, nil, error.details)
                }
            }
        }
    }
    
    @objc func generateThumbnail(_ call: CAPPluginCall) {
        guard let videoPath = call.getString("videoPath") else {
            call.reject("INVALID_OPTIONS", "videoPath is required")
            return
        }
        
        let timeAt = call.getDouble("timeAt") ?? 1.0 // 默认在第1秒生成缩略图
        let quality = call.getDouble("quality") ?? 0.8 // 默认压缩质量0.8
        
        // 处理路径兼容性：支持 file:// 开头的路径
        let actualVideoPath: String
        if videoPath.hasPrefix("file://") {
            actualVideoPath = String(videoPath.dropFirst(7)) // 移除 "file://" 前缀
        } else {
            actualVideoPath = videoPath
        }
        
        // 验证视频文件是否存在
        let videoURL = URL(fileURLWithPath: actualVideoPath)
        guard FileManager.default.fileExists(atPath: actualVideoPath) else {
            call.reject("FILE_NOT_FOUND", "Video file not found at path: \(actualVideoPath)")
            return
        }
        
        // 在后台线程生成缩略图
        DispatchQueue.global(qos: .userInitiated).async {
            let result = VideoRecorder.generateThumbnail(from: videoURL, timeAt: timeAt, quality: quality)
            
            DispatchQueue.main.async {
                switch result {
                case .success(let thumbnailPath):
                    call.resolve([
                        "thumbnailPath": thumbnailPath,
                        "videoPath": actualVideoPath,
                        "timeAt": timeAt,
                        "quality": quality
                    ])
                case .failure(let error):
                    call.reject(error.code, error.message, nil, error.details)
                }
            }
        }
    }



    // MARK: - Helper Methods

    private func createStopRecordingResponse(from result: StopRecordingResult) -> JSObject {
        return [
            "recordingId": result.recordingId,
            "videoPath": result.videoPath,
            "fileSize": NSNumber(value: result.fileSize),
            "duration": result.duration,
            "width": result.width,
            "height": result.height,
            "startTime": result.startTime,
            "endTime": result.endTime,
            "thumbnailPath": result.thumbnailPath ?? "",
            "mimeType": result.mimeType
        ]
    }

    private func createMediaFileDict(from result: StopRecordingResult) -> [String: Any] {
        return [
            "name": URL(fileURLWithPath: result.videoPath).lastPathComponent,
            "fullPath": result.videoPath,
            "type": result.mimeType,
            "size": result.fileSize,
            "lastModifiedDate": Date(timeIntervalSince1970: result.endTime / 1000).timeIntervalSince1970 * 1000
        ]
    }

    private func createMediaFileDict(from audioFile: AudioFile) -> [String: Any] {
        return [
            "name": audioFile.name,
            "fullPath": audioFile.path,
            "type": audioFile.mimeType,
            "size": audioFile.size,
            "lastModifiedDate": audioFile.lastModified.timeIntervalSince1970 * 1000
        ]
    }

    private func mapQualityFromNumber(_ quality: Int) -> VideoQuality {
        switch quality {
        case 0...25:
            return .low
        case 26...50:
            return .medium
        case 51...75:
            return .high
        default:
            return .highest
        }
    }

    private func permissionStateString(from status: AVAuthorizationStatus) -> String {
        switch status {
        case .authorized:
            return "granted"
        case .denied, .restricted:
            return "denied"
        case .notDetermined:
            return "prompt"
        @unknown default:
            return "prompt"
        }
    }

    private func photosPermissionStateString(from status: PHAuthorizationStatus) -> String {
        switch status {
        case .authorized:
            return "granted"
        case .denied, .restricted:
            return "denied"
        case .notDetermined:
            return "prompt"
        case .limited:
            return "granted"
        @unknown default:
            return "prompt"
        }
    }
}
